src/gc/roots.c

```diff
@@ -387,11 +387,16 @@ void MVM_gc_root_add_frame_registers_to_worklist(MVMThreadContext *tc, MVMGCWork
     MVMuint16  i, count, flag;
     MVMuint16 *type_map;
     MVMuint8  *flag_map;
-
     /* We only need to do any of this work if the frame is in dynamic scope. */
     if (frame->work) {
         /* Scan locals. */
-        if (frame->spesh_cand && frame->spesh_cand->local_types) {
+
+        MVMSpeshCandidate *spesh_cand = frame->spesh_cand;
+        MVMJitCode *jitcode = spesh_cand ? spesh_cand->jitcode : NULL;
+        if (jitcode && jitcode->local_types) {
+            type_map = jitcode->local_types;
+            count    = jitcode->num_locals;
+        } else if (frame->spesh_cand && frame->spesh_cand->local_types) {
```

Comment:
Not automatically a problem, but we do seem to have a proliferation of branches like this.